### PR TITLE
Add Gradients of MultivariateNormal.batch_log_pdf 

### PR DESCRIPTION
**Pull request by dwd31415**
This is a fix for #656

### CODE REVIEW

File: pyro/distributions/multivariate_normal.py [hunk 1]
```diff
@@ -29,15 +29,30 @@ def linear_solve_compat(matrix, matrix_chol, y):
 
 def matrix_inverse_compat(matrix, matrix_chol):
     """Computes the inverse of a positive semidefinite square matrix."""
-    assert matrix.requires_grad == matrix_chol.requires_grad
-    if matrix.requires_grad:
+    if matrix.requires_grad or matrix_chol.requires_grad:
```
Comment (fritzo) [inline pyro/distributions/multivariate_normal.py]: Could you help me understand under what conditions the assertion would fail?

File: tests/distributions/test_multivariate_normal.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        self.sample = Variable(torch.randn(N))
+
+        self.cholesky_mv_normalized = MultivariateNormal(self.mu, scale_tril=self.L, normalized=True)
+        self.cholesky_mv = MultivariateNormal(self.mu, scale_tril=self.L, normalized=False)
+
+        self.full_mv_normalized = MultivariateNormal(self.mu, self.sigma, normalized=True)
+        self.full_mv = MultivariateNormal(self.mu, self.sigma, normalized=False)
+
+    def test_log_pdf_gradients_cholesky(self):
+        self.cholesky_mv.log_pdf(self.sample).backward()
```
Comment (fritzo) [inline tests/distributions/test_multivariate_normal.py]: I would recommend using `torch.autograd.grad()` in these tests. We're finding it much more convenient than `.backward()`:
```py
from torch.autograd import grad
from tests.common import assert_equal

class TestMultivariateNormal(TestCase):
    ...
    def test_log_pdf_gradiennts_cholesky(self):
        grad1 = grad([self.cholesky_mv.log_pdf(self.sample)], [self.L])[0].data
        grad2 = grad([self.cholesky_mv_normalized.log_pdf(self.sample)], [self.L])[0].data
        assert_equal(grad1, grad2)
  ```

File: pyro/distributions/multivariate_normal.py [hunk 3]
```diff
@@ -29,15 +29,30 @@ def linear_solve_compat(matrix, matrix_chol, y):
...
         return potri_compat(matrix_chol)
 
 
+class _NormalizationConstant(Function):
+    @staticmethod
+    def forward(ctx, sigma, sigma_cholesky, inverse, dimension, return_zero):
+        ctx.save_for_backward(inverse, sigma_cholesky)
+        results = torch.Tensor(
+            [torch.log(sigma_cholesky.diag()).sum() + (dimension / 2) * np.log(2 * np.pi)]).type_as(sigma_cholesky)
+        return results if not return_zero else torch.zeros(1).type_as(sigma_cholesky)
```
Comment (fritzo) [inline pyro/distributions/multivariate_normal.py]: I would recommend (1) preserving the parameter name `normalized` throughout, and (2) avoiding unnecessary computation
```py
if not normalized:
    return torch.zeros(1).type_as(sigma_cholesky)
return sigma_cholesky.diag().log().sum() + ...  # no need for .type_as() here
```

Comment (dwd31415) [reply pyro/distributions/multivariate_normal.py]: I will split the if statement, but `return sigma_cholesky.diag().log().sum()` sadly does not work. The `.sum()` does not return a Tensor but a float (or whatever the type is) and pytorch won't accept that return type. That is the reason for the `torch.Tensor(...).type_as`.

File: pyro/distributions/multivariate_normal.py [hunk 4]
```diff
@@ -29,15 +29,30 @@ def linear_solve_compat(matrix, matrix_chol, y):
...
-    if matrix.requires_grad:
+    if matrix.requires_grad or matrix_chol.requires_grad:
         # If derivatives are required, use the more expensive inverse.
         return torch.inverse(matrix)
     else:
         # Use the cheaper Cholesky based potri.
         return potri_compat(matrix_chol)
 
 
+class _NormalizationConstant(Function):
```
Comment (fritzo) [inline pyro/distributions/multivariate_normal.py]: nit: It would be nice to have a one-line comment explaining the purpose of this class, something like:
> This computes either zero or the true normalization constant depending on `normalized`, but always computes the true tradient.

File: pyro/distributions/multivariate_normal.py (hunk 1 above)
Comment (dwd31415) [reply pyro/distributions/multivariate_normal.py]: The problem here is potrf_compat on PyTorch < 0.3.0: In that case `Variable(torch.potrf(sigma.data))` is returned and if `sigma.requires_grad` was true then they don't match up any more. The unit tests didn't catch this until now because they did not need any gradients.

File: pyro/distributions/multivariate_normal.py (hunk 3 above)
Comment (fritzo) [reply pyro/distributions/multivariate_normal.py]: Gah, I forgot about that `.sum()` behavior, but I'm not sure your version supports batching. What about this?
```py
sigma_cholesky.diag().log().sum(0) + ...
```

File: tests/distributions/test_multivariate_normal.py (hunk 2 above)
Comment (dwd31415) [reply tests/distributions/test_multivariate_normal.py]: Thanks, it is much cleaner like this.

File: pyro/distributions/multivariate_normal.py (hunk 3 above)
Comment (fritzo) [reply pyro/distributions/multivariate_normal.py]: Hmm, even `.diag()` doesn't support batching...

Comment (dwd31415) [reply pyro/distributions/multivariate_normal.py]: That seems to work.

Comment (dwd31415) [reply pyro/distributions/multivariate_normal.py]: Yeah, but this whole thing does not support batched parameters, so that does not really matter for now.
